### PR TITLE
BUG: groupby.describe on a frame with duplicate column names

### LINKED ISSUE #50806: BUG: groupby.describe on a frame with duplicate column names
xref #46944 

```
pd.set_option("display.max_columns", None)

df = DataFrame([[0, 1, 2, 3]])
df.columns = [0, 1, 2, 0]
gb = df.groupby(df[1])
print(gb.describe(percentiles=[]).to_string())
#       0                                                           2                        
#   count mean std  min  50%  max count mean std  min  50%  max count mean std  min  50%  max
# 1                                                                                          
# 1   1.0  0.0 NaN  0.0  0.0  0.0   1.0  3.0 NaN  3.0  3.0  3.0   1.0  2.0 NaN  2.0  2.0  2.0
```

With duplicate column names, describe only outputs the values for the first column. There should be two columns named `0` here. This issue occurs anywhere the groupby code is using `_selected_obj` within the `_group_selection_context`.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #50806 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

ASVs are below.

```
       before           after         ratio
     [a063af0e]       [185e4f8e]
     <groupby_select_obj_dup_cols~2>       <groupby_select_obj_dup_cols>
-        769±10μs          694±6μs     0.90  groupby.GroupByMethods.time_dtype_as_group('uint', 'first', 'transformation', 5)
-         758±3μs          683±9μs     0.90  groupby.GroupByMethods.time_dtype_as_group('int16', 'last', 'transformation', 5)
-     11.9±0.04ms       10.7±0.2ms     0.90  groupby.AggEngine.time_dataframe_cython(True)
-         846±4μs          759±6μs     0.90  groupby.GroupByMethods.time_dtype_as_group('float', 'min', 'transformation', 5)
-         857±5μs          769±3μs     0.90  groupby.GroupByMethods.time_dtype_as_group('float', 'max', 'transformation', 5)
-         758±3μs          678±2μs     0.89  groupby.GroupByMethods.time_dtype_as_group('datetime', 'last', 'transformation', 5)
-         763±3μs          682±8μs     0.89  groupby.GroupByMethods.time_dtype_as_group('datetime', 'first', 'transformation', 5)
-         766±1μs          683±4μs     0.89  groupby.GroupByMethods.time_dtype_as_group('datetime', 'max', 'transformation', 5)
-         775±2μs          691±7μs     0.89  groupby.GroupByMethods.time_dtype_as_group('int16', 'min', 'transformation', 5)
-         860±8μs          766±7μs     0.89  groupby.GroupByMethods.time_dtype_as_group('float', 'sum', 'transformation', 5)
-         775±3μs          689±8μs     0.89  groupby.GroupByMethods.time_dtype_as_group('int16', 'first', 'transformation', 5)
-         847±4μs          752±9μs     0.89  groupby.GroupByMethods.time_dtype_as_group('float', 'prod', 'transformation', 5)
-         759±2μs          674±2μs     0.89  groupby.GroupByMethods.time_dtype_as_group('uint', 'prod', 'transformation', 5)
-         709±3μs         627±10μs     0.88  groupby.GroupByMethods.time_dtype_as_group('object', 'last', 'transformation', 5)
-         765±2μs          676±4μs     0.88  groupby.GroupByMethods.time_dtype_as_group('int16', 'prod', 'transformation', 5)
-        854±10μs         754±10μs     0.88  groupby.GroupByMethods.time_dtype_as_group('float', 'first', 'transformation', 5)
-         766±4μs         675±20μs     0.88  groupby.GroupByMethods.time_dtype_as_group('uint', 'max', 'transformation', 5)
-         775±3μs          683±6μs     0.88  groupby.GroupByMethods.time_dtype_as_group('datetime', 'min', 'transformation', 5)
-         670±1μs          590±9μs     0.88  groupby.GroupByMethods.time_dtype_as_group('uint', 'cumcount', 'transformation', 5)
-         767±4μs         674±20μs     0.88  groupby.GroupByMethods.time_dtype_as_group('int16', 'sum', 'transformation', 5)
-         777±2μs         683±10μs     0.88  groupby.GroupByMethods.time_dtype_as_group('int', 'first', 'transformation', 5)
-         858±7μs          753±7μs     0.88  groupby.GroupByMethods.time_dtype_as_group('float', 'last', 'transformation', 5)
-         772±5μs          678±1μs     0.88  groupby.GroupByMethods.time_dtype_as_group('int', 'last', 'transformation', 5)
-         661±2μs          579±1μs     0.88  groupby.GroupByMethods.time_dtype_as_group('int16', 'cumcount', 'transformation', 5)
-         588±2μs          513±1μs     0.87  groupby.GroupByMethods.time_dtype_as_group('object', 'cumcount', 'transformation', 5)
-         779±3μs          680±9μs     0.87  groupby.GroupByMethods.time_dtype_as_group('int16', 'max', 'transformation', 5)
-         781±3μs         677±20μs     0.87  groupby.GroupByMethods.time_dtype_as_group('int', 'max', 'transformation', 5)
-        772±10μs          669±9μs     0.87  groupby.GroupByMethods.time_dtype_as_group('int', 'sum', 'transformation', 5)
-         663±3μs          574±9μs     0.87  groupby.GroupByMethods.time_dtype_as_group('int', 'cumcount', 'transformation', 5)
-         769±7μs         665±20μs     0.86  groupby.GroupByMethods.time_dtype_as_group('uint', 'sum', 'transformation', 5)
-         770±7μs         666±20μs     0.86  groupby.GroupByMethods.time_dtype_as_group('int', 'prod', 'transformation', 5)
-         645±2μs          557±1μs     0.86  groupby.GroupByMethods.time_dtype_as_group('datetime', 'cumcount', 'transformation', 5)
-         729±4μs         629±10μs     0.86  groupby.GroupByMethods.time_dtype_as_group('float', 'cumcount', 'transformation', 5)
-         779±2μs         673±20μs     0.86  groupby.GroupByMethods.time_dtype_as_group('uint', 'min', 'transformation', 5)
-         717±3μs         618±20μs     0.86  groupby.GroupByMethods.time_dtype_as_group('object', 'first', 'transformation', 5)
-         781±2μs         665±20μs     0.85  groupby.GroupByMethods.time_dtype_as_group('int', 'min', 'transformation', 5)
-        782±20μs         665±20μs     0.85  groupby.GroupByMethods.time_dtype_as_group('uint', 'last', 'transformation', 5)
-         580±6μs          491±7μs     0.85  groupby.TransformNaN.time_first
-         448±2μs         353±10μs     0.79  groupby.SumBools.time_groupby_sum_booleans
-         374±1μs          292±1μs     0.78  groupby.GroupByMethods.time_dtype_as_group('int', 'cumcount', 'direct', 5)
-         372±3μs          284±2μs     0.76  groupby.GroupByMethods.time_dtype_as_group('datetime', 'cumcount', 'direct', 5)
-         358±3μs          272±1μs     0.76  groupby.GroupByMethods.time_dtype_as_group('object', 'cumcount', 'direct', 5)
-         376±7μs        283±0.7μs     0.75  groupby.GroupByMethods.time_dtype_as_group('uint', 'cumcount', 'direct', 5)
-         374±2μs          279±3μs     0.75  groupby.GroupByMethods.time_dtype_as_group('float', 'cumcount', 'direct', 5)
-         382±7μs          283±7μs     0.74  groupby.GroupByMethods.time_dtype_as_group('int16', 'cumcount', 'direct', 5)
-         181±2μs         85.9±1μs     0.47  groupby.GroupByMethods.time_dtype_as_group('float', 'sum', 'direct', 5)
-         175±3μs       82.9±0.3μs     0.47  groupby.GroupByMethods.time_dtype_as_group('int', 'sum', 'direct', 5)
-         175±2μs       82.9±0.2μs     0.47  groupby.GroupByMethods.time_dtype_as_group('int16', 'first', 'direct', 5)
-       178±0.2μs       84.1±0.9μs     0.47  groupby.GroupByMethods.time_dtype_as_group('int', 'first', 'direct', 5)
-         180±3μs         83.8±2μs     0.47  groupby.GroupByMethods.time_dtype_as_group('float', 'max', 'direct', 5)
-       177±0.9μs       82.3±0.2μs     0.47  groupby.GroupByMethods.time_dtype_as_group('uint', 'min', 'direct', 5)
-         175±1μs         81.4±1μs     0.47  groupby.GroupByMethods.time_dtype_as_group('int16', 'sum', 'direct', 5)
-         179±1μs       83.2±0.9μs     0.46  groupby.GroupByMethods.time_dtype_as_group('int16', 'max', 'direct', 5)
-         179±3μs         82.7±1μs     0.46  groupby.GroupByMethods.time_dtype_as_group('uint', 'max', 'direct', 5)
-         181±2μs       83.6±0.6μs     0.46  groupby.GroupByMethods.time_dtype_as_group('float', 'min', 'direct', 5)
-         180±1μs       82.9±0.6μs     0.46  groupby.GroupByMethods.time_dtype_as_group('uint', 'first', 'direct', 5)
-         178±1μs       81.9±0.3μs     0.46  groupby.GroupByMethods.time_dtype_as_group('int', 'min', 'direct', 5)
-         178±1μs       81.4±0.7μs     0.46  groupby.GroupByMethods.time_dtype_as_group('float', 'last', 'direct', 5)
-         177±1μs       81.3±0.5μs     0.46  groupby.GroupByMethods.time_dtype_as_group('uint', 'sum', 'direct', 5)
-         180±2μs       82.4±0.1μs     0.46  groupby.GroupByMethods.time_dtype_as_group('int', 'max', 'direct', 5)
-         176±1μs         80.6±2μs     0.46  groupby.GroupByMethods.time_dtype_as_group('datetime', 'max', 'direct', 5)
-       170±0.4μs         77.5±1μs     0.46  groupby.GroupByMethods.time_dtype_as_group('datetime', 'last', 'direct', 5)
-         175±3μs       79.7±0.2μs     0.46  groupby.GroupByMethods.time_dtype_as_group('float', 'prod', 'direct', 5)
-         179±1μs       81.5±0.3μs     0.46  groupby.GroupByMethods.time_dtype_as_group('int16', 'min', 'direct', 5)
-         179±4μs         81.5±2μs     0.45  groupby.GroupByMethods.time_dtype_as_group('float', 'first', 'direct', 5)
-       174±0.5μs         79.0±3μs     0.45  groupby.GroupByMethods.time_dtype_as_group('datetime', 'first', 'direct', 5)
-         175±1μs         79.4±2μs     0.45  groupby.GroupByMethods.time_dtype_as_group('datetime', 'min', 'direct', 5)
-         172±2μs         77.2±1μs     0.45  groupby.GroupByMethods.time_dtype_as_group('int', 'last', 'direct', 5)
-         170±2μs       75.6±0.3μs     0.44  groupby.GroupByMethods.time_dtype_as_group('uint', 'prod', 'direct', 5)
-       170±0.9μs       75.5±0.3μs     0.44  groupby.GroupByMethods.time_dtype_as_group('int', 'prod', 'direct', 5)
-         170±2μs       75.5±0.7μs     0.44  groupby.GroupByMethods.time_dtype_as_group('int16', 'prod', 'direct', 5)
-         172±4μs       76.2±0.2μs     0.44  groupby.GroupByMethods.time_dtype_as_group('int16', 'last', 'direct', 5)
-       175±0.6μs         75.7±2μs     0.43  groupby.GroupByMethods.time_dtype_as_group('uint', 'last', 'direct', 5)
-         158±1μs         66.0±2μs     0.42  groupby.GroupByMethods.time_dtype_as_group('object', 'first', 'direct', 5)
-         160±2μs         65.5±1μs     0.41  groupby.GroupByMethods.time_dtype_as_group('object', 'last', 'direct', 5)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1024,6 +1026,9 @@ def _set_group_selection(self) -> None:
             # GH12839 clear selected obj cache when group selection changes
             ax = self.obj._info_axis
             self._group_selection = ax.difference(Index(groupers), sort=False).tolist()
+            self._group_selection = [
+                idx for idx, label in enumerate(ax) if label not in groupers
+            ]
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:1030]: I timed this using the code below, this is faster on DataFrames with < 4000 columns, but becomes significantly slower on larger. Going to also post ASVs. Not sure if there is a more efficient method.

Top line is main, bottom line is this PR

```
10
71.8 µs ± 120 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
1.18 µs ± 2.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

100
108 µs ± 428 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
6.47 µs ± 19.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

1000
113 µs ± 2.24 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
63.1 µs ± 238 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
112 µs ± 141 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
129 µs ± 421 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

4000
115 µs ± 129 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
260 µs ± 907 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

8000
127 µs ± 137 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
511 µs ± 1.99 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

10000
132 µs ± 2.66 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
648 µs ± 2.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100000
349 µs ± 265 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
7.2 ms ± 16.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Code:

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=5))
    print(size)
    %timeit ax.difference(Index(groupers), sort=False).tolist()
    %timeit [idx for idx, label in enumerate(ax) if label not in groupers]
    print()
```

Comment (phofl) [reply pandas/core/groupby/groupby.py:1030]: This should be a bit faster: 

```
    y = ax.get_indexer_for(groupers)
    z = np.delete(np.arange(len(ax)), y)
```

Comment (phofl) [reply pandas/core/groupby/groupby.py:1030]: Another thing: Can't we just drop the duplicates before selecting the columns?

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1030]: Thanks! Still slower than the original, but a definite improvement. Also, the original is doing a different computation, so not good to compare. I would guess typical usage is with a DataFrame with < 100 columns, so it seems beneficial to have the list-comprehension version as well.

> Another thing: Can't we just drop the duplicates before selecting the columns?

We want to keep the duplicates. If the columns (ignoring the grouping) are `['a', 'b', 'a']`, we want to persist these in `DataFrameGroupBy._selected_obj` and in that order. This is why I'm switching from labels to ilocs.

These timings are on my laptop, so a bit slower than the ones posted previously. The third row is @phofl's alternative.

```
10
215 µs ± 9.12 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
2.74 µs ± 16.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
299 µs ± 175 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100
228 µs ± 5.65 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
15.2 µs ± 51 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
118 µs ± 1.12 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

1000
226 µs ± 2.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
155 µs ± 630 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
175 µs ± 1.42 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
241 µs ± 3.22 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
291 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
246 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

4000
239 µs ± 498 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
655 µs ± 1.65 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
395 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

8000
276 µs ± 5.05 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
1.21 ms ± 1.18 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
652 µs ± 1.55 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

10000
264 µs ± 2.33 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
1.64 ms ± 2.78 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
822 µs ± 8.11 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

100000
619 µs ± 2.72 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
18.2 ms ± 1.77 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.6 ms ± 1.36 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Code:

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=20))
    print(size)
    %timeit ax.difference(Index(groupers), sort=False).tolist()
    %timeit np.array(idx for idx, label in enumerate(ax) if label not in groupers)
    %timeit np.delete(np.arange(len(ax)), ax.get_indexer_for(list(groupers)))
    print()
```

Comment (phofl) [reply pandas/core/groupby/groupby.py]: Sorry I meant removing duplicates in _group_selection. They will persist when you do self.obj[self._group_selection] but no further duplication when _group_selection is unique. The order is changed though

Comment (phofl) [reply pandas/core/groupby/groupby.py]: How did you get the list comprehension so fast? Looks a bit odd that it’s constant

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: I do think we need order to come out the same as the input.

The timings for the list comprehension are the time it takes for the first iteration to fail 🤦 will fix and repost.

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: @phofl - timings have been updated.

**Comment by rhshadrach** [comment]
@mroeschke - friendly ping. What do you think about the conditional logic based on the number of columns here?

File: pandas/tests/groupby/test_groupby_dropna.py [hunk 2]
```diff
@@ -578,6 +581,8 @@ def test_categorical_reducers(
     result = getattr(gb_keepna, reduction_func)(*args)
 
     # size will return a Series, others are DataFrame
+    print(result.index.dtype)
```
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby_dropna.py:584]: Guessing these `print`s are not needed

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -1023,7 +1025,14 @@ def _set_group_selection(self) -> None:
         if len(groupers):
             # GH12839 clear selected obj cache when group selection changes
             ax = self.obj._info_axis
-            self._group_selection = ax.difference(Index(groupers), sort=False).tolist()
+            if len(ax) < 1000:
```
Comment (mroeschke) [inline pandas/core/groupby/groupby.py:1028]: Generally I don't like constant thresholds like this as it's not guaranteed to hold overtime and is susceptible to behavior difference that are data size related (which can be hard to debug).

I would prefer just to use one path.

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: New implementation: `np.arange(len(ax))[~ax.isin(groupers)]`. Still not as fast as Python lists comprehension for small sizes, but faster than previous. This is faster than main (and now correct), and I think this is quick enough for this portion of the code. For example, for a 10 column 1000 row frame for floats I'm seeing `df.groupby('a').sum()` take 300 microseconds, 20 of which we're spending for this computation.

<details>
<summary>Timings</summary>

```
10
1.84 µs ± 45.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
48.7 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
19.7 µs ± 844 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

100
10.2 µs ± 24.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
50.8 µs ± 121 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
19.6 µs ± 82.6 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

1000
102 µs ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
85.1 µs ± 1.82 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
23.4 µs ± 105 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

2000
200 µs ± 811 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
115 µs ± 1.2 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
27.2 µs ± 110 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

4000
395 µs ± 675 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
190 µs ± 6.03 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
37 µs ± 454 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

8000
808 µs ± 963 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
329 µs ± 202 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
52 µs ± 195 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

10000
1.02 ms ± 22.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
399 µs ± 932 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
69.5 µs ± 4.53 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

100000
10.8 ms ± 621 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
3.58 ms ± 68.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
1.1 ms ± 17.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

</details>

<details>
<summary>Code</summary>

```
for size in (10, 100, 1000, 2000, 4000, 8000, 10000, 100000):
    ax = pd.Index(np.random.randint(0, 100, size=size))
    groupers = frozenset(np.random.randint(0, 100, size=5))
    print(size)
    %timeit np.array([idx for idx, label in enumerate(ax) if label not in groupers])
    %timeit np.delete(np.arange(len(ax)), ax.get_indexer_for(list(groupers)))
    %timeit np.arange(len(ax))[~ax.isin(groupers)]
    print()
```

</details>

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: Thanks - I found a more efficient implementation, timings are in https://github.com/pandas-dev/pandas/pull/50846#discussion_r1087198598

File: pandas/tests/groupby/test_groupby_dropna.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/tests/groupby/test_groupby_dropna.py]: Doh - thanks

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/groupby/groupby.py]: Awesome!

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -726,7 +726,9 @@ def _selected_obj(self):
 
         if self._selection is None or isinstance(self.obj, Series):
             if self._group_selection is not None:
-                return self.obj[self._group_selection]
+                return self.obj._take(
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:729]: I think this is the same as _obj_with_exclusions, which should already be cached, so we could avoid making a copy

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: This is great - not only that, but we can also avoid all the code that determines `_group_selection`. I've turned `_group_selection` into a Boolean flag.

Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py]: should we expect this change to affect the timings in the OP?

File: pandas/tests/groupby/test_function.py [hunk 5]
```diff
@@ -1594,3 +1615,29 @@ def test_multiindex_group_all_columns_when_empty(groupby_func):
...
+        msg = "GH#50845 - corrwith fails when there are duplicate columns"
+        request.node.add_marker(pytest.mark.xfail(reason=msg))
+    df = DataFrame([[1, 3, 6], [1, 4, 7], [2, 5, 8]], columns=list("abb"))
+    args = get_groupby_method_args(groupby_func, df)
+    gb = df.groupby("a", as_index=as_index)
+    result = getattr(gb, groupby_func)(*args)
+
+    if groupby_func in ("size", "ngroup", "cumcount"):
+        expected = getattr(
+            df.take([0, 1], axis=1).groupby("a", as_index=as_index), groupby_func
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_function.py:1632]: nitpick: can you avoid chaining take/gropby/getattr here (and in L1639)?  easier to grok if something goes wrong

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: ASVs updated; essentially the same results.

File: pandas/tests/groupby/test_function.py (hunk 5 above)
Comment (rhshadrach) [reply pandas/tests/groupby/test_function.py]: Done.

**Comment by rhshadrach** [comment]
@mroeschke @jbrockmendel - friendly ping.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Pull request merged**